### PR TITLE
Enable copy paste from Datatable

### PR DESCRIPTION
**Pull request by alhoo**
Add support for copy pasting data from datatables

Very simple implementation of copy pasting. Probably needs some more work and no unit tests with this commit, but hopefully gives a quick start to finalizing the rest. Fixes issue #7762.

### CODE REVIEW

**Comment by alhoo** [comment]
The resulting paste is not ordered correctly and/but the paste also contains an additional column with the original line number :D

**Comment by alhoo** [comment]
Apparently the slickgrid copy manager captures/modifies some click events and causes some of the current unit tests to break (atleast tests/integration/widgets/tables/test_cell_editors.py::Test_*Editor::test_editing_updates_source).

With this copy-paste-plugin I need to use a triple-click with the mouse to edit the values (tested manually with firefox and chrome) where the current master only needs 2 clicks. We could either accept the new behaviour and change our unit tests until the plugin gets fixed by slickgrid or then the slickgrid plugin needs to be updated. Slickgrid has atleast the following known issues which might be related: #1101 and #994.

As for the incorrectly ordered rows: the copy manager seems to get the badly ordered list from the selection tool so the problem might be in the RowSelectionModel, so once again not a problem in the Bokeh code.

**Comment by alhoo** [comment]
I got to agree with @bryevdv that the changed cell editing logic is annoying to use. I disabled the copy-paste feature from editable datatables, which is also a bit annoying but I guess it is less annoying for the average user.

I guess some users would like to have the copy-paste also with editable fields even if that breaks/changes the editing gesture. Maybe we could add a [force] copy_paste=True [and change editing gesture... sorry about that] -flag for that?

**Comment by bryevdv** [comment]
@alhoo we consider putting it under a feature flag. My main concern is that when/if it is fixed upstream, we would be left with a no-longer-useful flag that would need to be kept around indefinitely deprecated in order to no break anyone's code. Before doing that, I would like to have at least engaged minimally with the devs working on the fork, to make sure there is not some simple fix that could be applied, or that it's not something they plan to imminently work on. Can you open up an issue on the issue tracker for the 6pac fork to ask about this? (Please link back to this issue)

**Comment by alhoo** [comment]
@bryevdv I tried to replicate the problem with slickgrid, but apparently the problem does not occur in the 6pac.github.io (master).

**Comment by alhoo** [comment]
I made some more testing and the triple-click problem seems to be fixed starting with the SlickGrid commit 9463a06 (release v2.3.18 - update github.io).

File: bokehjs/src/lib/models/widgets/tables/data_table.ts [hunk 1]
```diff
@@ -255,6 +256,21 @@ export class DataTableView extends WidgetView {
...
+        const pluginOptions = {
+          dataItemColumnValueExtractor: function(val: any, col: any) {
+            let value: any = val[col.field]
+            if ( typeof value === 'string'){
+              value = value.replace(/\n/g, "\\n")
+            }
+            return value
+          },
+          includeHeaderWhenCopying : false,
+        }
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/tables/data_table.ts]: This code needs a cleanup. Please fix whitespace around `(`, `{` and `:`. `typeof` should be replaced with `isString()` function (from `core/util/types`). `any` types need to be replaced with more specific types or, if there is no other better choice, there needs to be a comment explaining why `any` is being used, including the best approximation of a possible type.

Comment (alhoo) [reply bokehjs/src/lib/models/widgets/tables/data_table.ts]: @mattpap I was unable get any output from tslint, and py.test -m codebase doesn't seem to test the typescript syntax. I could not find instructions on the "Getting Set Up" -guide. I tried to fix the issues by guessing what exactly you meant. Hopefully that is close enough.

Comment (alhoo) [reply bokehjs/src/lib/models/widgets/tables/data_table.ts]: Argh. Line 260 still needs a space before `{`.

**Comment by bryevdv** [comment]
> I made some more testing and the triple-click problem seems to be fixed starting with the SlickGrid commit 9463a06 (release v2.3.18 - update github.io).

That's good news, I will try updating our fork but FYI it will probably not be until late this week before I can get to it.

**Comment by bryevdv** [comment]
OK just as a quick test I tried to merge the upstream master and as expected our changes do not apply cleanly. @mattpap is probably the best person to take a look at this, since he made the changes to our fork. Ideally we can reduce/eliminate the differences between our fork and the upstream.

**Comment by alhoo** [comment]
I found the [problematic line](https://github.com/alhoo/SlickGrid/commit/361041e0409e445d3bcc7d5393ed030a40471c71) from the bokeh/slickgrid, but it would be nice to get the full upgrade to slickgrid.

**Comment by mattpap** [comment]
>  @mattpap is probably the best person to take a look at this

Yeah, but not earlier than the layout is done.

**Comment by bryevdv** [comment]
@alhoo this is pretty fantastic. Can you comment on how far `d193f8a` is from the current HEAD of `6pac/slickgrid`? It looks like you have consolidated all the commits from our for into one patch, would it be possible to split them up? Currently they are also failing out linter due to trailing whitespace. If that's an issue we could exclude the patch file. 

@mattpap can you comment? It looks like the layout PR has almost no changes to anything data table related, so it seems like this work would not conflict. I'd be interested to get this feature in and turn out a 1.0.4 release next week. 

Are our changes something that the upstream might consider if we made a PR?

**Comment by alhoo** [comment]
Oh and d193f8a is the current HEAD.

**Comment by bryevdv** [comment]
@alhoo I have merged the PR in `bokeh/SlickGrid` can you update`package.json` to build from the new head, and do some minimal testing, then submit a new PR?

**Pull request closed without merging**